kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/InBandDecryptionManager.java

```diff
@@ -207,36 +144,35 @@ private CompletionStage<List<DecryptState<E>>> resolveAll(String topicName,
             var decryptionVersion = decryptionVersion(topicName, partition, record);
             if (decryptionVersion != null) {
                 ByteBuffer wrapper = record.value();
-                cacheKeys.add(decryptionVersion.wrapperVersion().readSpecAndEdek(wrapper, serde, CacheKey::new));
+                cacheKeys.add(decryptionVersion.wrapperVersion().readSpecAndEdek(wrapper, serde, DecryptionDekCache.CacheKey::new));
                 states.add(new DecryptState<>(decryptionVersion));
             }
             else {
                 // It's not encrypted, so use sentinels
-                cacheKeys.add(CacheKey.none());
+                cacheKeys.add(DecryptionDekCache.CacheKey.none());
                 states.add(DecryptState.none());
             }
         });
         // Lookup the decryptors for the cache keys
-        return filterThreadExecutor.completingOnFilterThread(decryptorCache.getAll(cacheKeys))
+        return filterThreadExecutor.completingOnFilterThread(dekCache.getAll(cacheKeys, filterThreadExecutor))
```

Comment:
I don't like how the `CacheKey` needs to be known about by the decryption manager. It's worth talking about why the `CacheKey` even needs to exist... The cipher spec needs be passed around alongside the edek because the signature of the `DekManager` requires a `CipherSpec` in order to issue a `Dek`. I made that decision when writing `DekManager` because I wanted a guarantee that we couldn't possible use the same key material to do encryption with different ciphers. (Why? Because having two different ciphertexts encrypted with the same key could possible aid cryptanalysis). But really this is only a requirement for encryption. For decryption we could just say "Gimme a `Dek` for this encrypted DEK" and delay requiring the `CipherSpec` until we actually try to use the `Dek` to decrypt something.

If we made a change like that then the cache key could just be the edek. 

I'm not really sure it's worth doing this right now, but thought I'd at least mention this.

Reply:
Having the same cache key for encrypt/decrypt might make refactoring them to share code tidier. I could imagine them becoming a `DekCache` with composed loader/eviction functions and the ability to offer the `key -> value` to another `DekCache`.

Reply:
> The cipher spec needs be passed around alongside the edek because the signature of the DekManager requires a CipherSpec in order to issue a Dek

That feels like a reasonable pairing, so should we actually update the model to join them explicitly rather than as separate args or hiding the relationship in the CacheKey? It feels to me a bit like the same conversation we had a few weeks back around wrapping Kek in a type and stopping the generics spreading throughout the filter.  

I think the proposed solution in this PR is good enough to merge. We can revisit at a later date as part of larger review?

Reply:
> Having the same cache key for encrypt/decrypt might make refactoring them to share code tidier.

How would that work? On the encrypt path the key is the `{kekId, cipherSpec}` and on the decrypt path it's the `{edek,  cipherSpec}`. Taking the union, `{kekId, cipherSpec, edek}` means we can't easily evict an old DEK on the encryption path, because we don't know the edek. 

> That feels like a reasonable pairing

It's more _symmetrical_ if the cipherSpec is required for both the `generateDek()` and the `decryptEdek()` operations on `Dek` manager. But the point I was trying to make is that it's not _required_.

It should be part of `generateDek()` in order to guarantee that the same kek material cannot be used with different ciphers, but once you have that guarantee then you don't really need it to appear explicitly in the signature of `decryptEdek()`; it would be enough to pass it to `Dek.decryptor()`. And if we made that change then it doesn't need to be in the `CacheKey` of the decryption cache. It would also simplify the deserialization of the wrapper, because we could delay deserializing it, rather than needing it in both "phases" of the deserialization. Neither of these really feels like a brilliant reason to make the change, but the desirabilty of symmetry doesn't feel like a great reason to keep things the way they are either.